std/traits.d

```diff
@@ -5548,6 +5549,132 @@ template BuiltinTypeOf(T)
     else                                        static assert(0);
 }
 
+/**
+    A template designed to function in the same manner as the
+    built-in `typeof` utility. As it is a template, `Typeof` can be
+    passed around like any other symbol, unlike `typeof`. Also,
+    whereas `typeof` will not accept symbols or expressions that
+    evaluate to a type (e.g., `typeof(int)` will not compile),
+    `Typeof` will accept both values and types.
+
+    Params:
+        T =  The symbol or expression to take the type of. Can be either a type
+             or a value. If it is something that does not have a valid
+             type (such as a module name or an expression that
+             evaluates to the special `__error__` type), an error
+             will be raised at compile time.
+
+    Returns:
+        The type of the given symbol or expression.
+
+    Note:
+        If a template is passed to `Typeof`, the result type will be `void`.
+*/
+template Typeof(T...)
+if (T.length == 1)
+{
+    static if (__traits(isTemplate, T[0]))
+        alias Typeof = void;
+    else static if (is(T[0] TypeOfT0))
+        alias Typeof = TypeOfT0;
+    else static if (is(typeof(T[0]) TypeOfT0))
+        alias Typeof = TypeOfT0;
+    else
+        static assert(false, "Cannot take the type of " ~ T[0].stringof);
+}
+
+///
+@safe unittest
```

Comment:
I'd add `pure` and convert the `assert`s into `static assert`s.

Reply:
I don't see a point in adding pure as `Typeof` is a template, not a function.

Do we have a policy on using `static assert` vs. `assert` in unit tests for stuff that can be checked at compile time? I don't know if it's necessarily a good idea for _compilation_ to fail when building with tests enabled, but that's just a vague sense of unease on my part.

Reply:
static assert is better than assert but keep in in the unittest so it doesn't slow down regular compilation

Reply:
I've changed all asserts to static asserts.